core/bytes/bytes.odin

```diff
@@ -295,22 +297,41 @@ split_after_iterator :: proc(s: ^[]byte, sep: []byte) -> ([]byte, bool) {
 
 
 index_byte :: proc(s: []byte, c: byte) -> int {
-	for i := 0; i < len(s); i += 1 {
-		if s[i] == c {
-			return i
+	_index_byte :: #force_inline proc "contextless" (s: []byte, c: byte) -> int {
+		for ch, i in s {
+			if ch == c {
+				return i
+			}
 		}
+		return -1
+	}
+
+	// NOTE(Feoramund): On my Alder Lake CPU, I have only witnessed a
```

Comment:
The main reason for performance regression without optimization in the old version of the code would be the lack of the SROA pass negatively interacting with how `core:simd/x86` is implemented (`#force_inline` at the IR level leaves a lot of redundant loads/stores that the pass removes, which adds up when every intrinsic is a wrapper).

This should not be the case with `core:simd`since those are compiler builtins that are emitted directly.  Did you re-benchmark with various optimization levels?

Invoking the slow path `when ODIN_ARCH == .wasm64p32 || ODIN_ARCH == .wasm32` AND `!intrinsics.has_target_feature("simd128")` is also probably better.

Reply:
Indeed re-running the `benchmark/simd/util` with `odin test . -o:none`, the SIMD version dramatically out-performs the plain version for larger sizes.

simd (cold):
```
        +++      15B | 110ns
 (last) +++      15B | 100ns
        +++      16B | 100ns
 (last) +++      16B | 100ns
        +++      17B | 71ns
 (last) +++      17B | 81ns
        +++      31B | 100ns
 (last) +++      31B | 110ns
        +++      32B | 201ns
 (last) +++      32B | 170ns
        +++      33B | 120ns
 (last) +++      33B | 140ns
        +++     256B | 210ns
 (last) +++     256B | 181ns
        +++     512B | 170ns
 (last) +++     512B | 170ns
        +++  1.00KiB | 241ns
 (last) +++  1.00KiB | 230ns
        +++  1.00MiB | 99.737µs
 (last) +++  1.00MiB | 84.047µs
        +++  1.00GiB | 90.245121ms
 (last) +++  1.00GiB | 92.246174ms
 ```
 
plain (cold):
```
        +++      15B | 90ns
 (last) +++      15B | 91ns
        +++      16B | 60ns
 (last) +++      16B | 60ns
        +++      17B | 50ns
 (last) +++      17B | 70ns
        +++      31B | 71ns
 (last) +++      31B | 101ns
        +++      32B | 71ns
 (last) +++      32B | 80ns
        +++      33B | 71ns
 (last) +++      33B | 81ns
        +++     256B | 301ns
 (last) +++     256B | 310ns
        +++     512B | 571ns
 (last) +++     512B | 621ns
        +++  1.00KiB | 1.102µs
 (last) +++  1.00KiB | 1.182µs
        +++  1.00MiB | 1.063621ms
 (last) +++  1.00MiB | 1.167035ms
        +++  1.00GiB | 986.739524ms
 (last) +++  1.00GiB | 1.216933171s
```

nb: The "plain is faster than simd for small messages" largely/entirely goes away with `-o:minimal`/`-o:speed` .

Reply:
> This should not be the case with core:simdsince those are compiler builtins that are emitted directly. Did you re-benchmark with various optimization levels?
That's an old comment. I re-benchmarked a little while after that commit and saw that there was no longer a speed decrease with the cross-platform SIMD implementation. I planned on merging all of this into `core:bytes` after seeing that.

Reply:
This is the change I made.
```cpp
@private SCAN_WIDTH :: 8 * size_of(uintptr)

when SCAN_WIDTH == 64 {
	@(private, rodata)
	simd_scanner_indices := #simd[SCAN_WIDTH]u8 {
		 0,  1,  2,  3,  4,  5,  6,  7,
		 8,  9, 10, 11, 12, 13, 14, 15,
		16, 17, 18, 19, 20, 21, 22, 23,
		24, 25, 26, 27, 28, 29, 30, 31,
		32, 33, 34, 35, 36, 37, 38, 39,
		40, 41, 42, 43, 44, 45, 46, 47,
		48, 49, 50, 51, 52, 53, 54, 55,
		56, 57, 58, 59, 60, 61, 62, 63,
	}
} else {
	@(private, rodata)
	simd_scanner_indices := #simd[SCAN_WIDTH]u8 {
		 0,  1,  2,  3,  4,  5,  6,  7,
		 8,  9, 10, 11, 12, 13, 14, 15,
		16, 17, 18, 19, 20, 21, 22, 23,
		24, 25, 26, 27, 28, 29, 30, 31,
	}
}
```